// @aws-cdk/aws-ec2/test/integ.instance-init.js
#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const path = require("path");
const cdk = require("@aws-cdk/core");
const ec2 = require("../lib");
const app = new cdk.App();
const stack = new cdk.Stack(app, 'integ-init', {
    env: {
        account: process.env.CDK_INTEG_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
        region: process.env.CDK_INTEG_REGION || process.env.CDK_DEFAULT_REGION,
    },
});
const vpc = ec2.Vpc.fromLookup(stack, 'VPC', { isDefault: true });
const tmpDir = fs.mkdtempSync('/tmp/cfn-init-test');
fs.writeFileSync(path.resolve(tmpDir, 'testFile'), 'Hello World!\n');
new ec2.Instance(stack, 'Instance2', {
    vpc,
    vpcSubnets: { subnetType: ec2.SubnetType.PUBLIC },
    instanceType: ec2.InstanceType.of(ec2.InstanceClass.BURSTABLE2, ec2.InstanceSize.MICRO),
    machineImage: new ec2.AmazonLinuxImage(),
    userDataCausesReplacement: true,
    initOptions: {
        timeout: cdk.Duration.minutes(30),
    },
    init: ec2.CloudFormationInit.fromConfigSets({
        configSets: {
            default: ['yumPreinstall', 'config'],
        },
        configs: {
            yumPreinstall: new ec2.InitConfig([
                ec2.InitPackage.yum('git'),
            ]),
            config: new ec2.InitConfig([
                ec2.InitFile.fromObject('/tmp/file2', {
                    stackId: stack.stackId,
                    stackName: stack.stackName,
                    region: stack.region,
                }),
                ec2.InitGroup.fromName('group1'),
                ec2.InitGroup.fromName('group2', 42),
                ec2.InitUser.fromName('sysuser1', {
                    groups: ['group1', 'group2'],
                    homeDir: '/home/sysuser1-custom',
                }),
                ec2.InitUser.fromName('sysuser2'),
                ec2.InitSource.fromAsset('/tmp/sourceDir', tmpDir),
            ]),
        },
    }),
});
app.synth();
